### PR TITLE
WIP: simplifying gat

### PR DESCRIPTION
**Pull request by kingjr**
Following https://github.com/mne-tools/mne-python/issues/1964,  I propose to
*    Drop the 'predict_type' option. If users want to use a decision_function or a predict_proba (or anything else we currently don't support) they could easily make their own classifier with a inherited class: e.g.

```
from sklearn.svm import SVC
class my_SVC(SVC):
    def predict(self, x):
        probas = super(my_SVC, self).predict_proba(x)
        return probas[:,0]
```

*  Drop the automatic definition of the scorer (AUC if predict_type is probabilistic, Accuracy if discrete, Least Square if Regression etc...) and let the user takes care of this by defining the scorer him/herself.

### CODE REVIEW

File: mne/decoding/time_gen.py [hunk 1]
```diff
@@ -411,10 +396,12 @@ def score(self, epochs=None, y=None, scorer=None, test_times=None):
...
-        from sklearn.metrics import (roc_auc_score, accuracy_score,
-                                     mean_squared_error)
-        from sklearn.base import is_classifier
-        from sklearn.preprocessing import LabelEncoder
+        from sklearn.metrics import accuracy_score
+
+        # Check scorer
+        if scorer is None:
+            scorer = accuracy_score
+        self.scorer = scorer
```
Comment (dengemann) [inline mne/decoding/time_gen.py]: If this attribute is not set at construction time it should end with an underscore.

File: mne/decoding/time_gen.py [hunk 2]
```diff
@@ -98,15 +95,6 @@ class GeneralizationAcrossTime(object):
...
         If None, empty dict. Defaults to None.
-    predict_type : {'predict', 'predict_proba', 'decision_function'}
-        Indicates the type of prediction:
-            'predict' : generates a categorical estimate of each trial.
-
-            'predict_proba' : generates a probabilistic estimate of each trial.
-
-            'decision_function' : generates a continuous non-probabilistic
-                estimate of each trial.
-        Default: 'predict'
```
Comment (dengemann) [inline mne/decoding/time_gen.py]: not sure, predict proba is a more common use case, not? We should directly support it via the API in a dumb non-expert way.

Comment (kingjr) [inline mne/decoding/time_gen.py]: I plead in favor of non discretizing clf outputs to maximize information but the most common case is actually SVC + accuracy.

**Comment by kingjr** [comment]
@dengemann Ok, so, following our discussion and to keep track, what we mostly need is a API that can guesses which scorer to use from the classifier implemented. Perhaps sklearn has this?

File: mne/decoding/time_gen.py (hunk 2 above)
Comment (dengemann) [inline mne/decoding/time_gen.py]: ok, but let's find an API for chosing between probability and discrete output, maybe based on the scoring.

File: mne/decoding/time_gen.py (hunk 1 above)
Comment (kingjr) [inline mne/decoding/time_gen.py]: I hesitated because it's gat.score(scorer=my_scorer). You confirm that we keep it as self.scorer_ in such case?

Comment (dengemann) [inline mne/decoding/time_gen.py]: Yes, everything that is not set at construction time needs this `_`.  And then you need to docment it in the constructor docs, telling that this attribute may be set when calling certain methods, etc.

Comment (kingjr) [inline mne/decoding/time_gen.py]: ok thanks!

File: mne/decoding/time_gen.py (hunk 2 above)
Comment (kingjr) [inline mne/decoding/time_gen.py]: I haven't found it.

Also, when is the next release?... Just to have an idea.

Comment (Eric89GXL) [inline mne/decoding/time_gen.py]: Shooting for Monday

Comment (kingjr) [inline mne/decoding/time_gen.py]: > Shooting for Monday
Damn, that's short

@cmoutard, do you have time to review it?

Comment (cmoutard) [inline mne/decoding/time_gen.py]: Ok in the coming days @kingjr

**Comment by agramfort** [comment]
no time to look in details now but if the example stays simple (as it should since it should cover 80% of use cases) and I see a lot of red lines I am happy.

**Comment by agramfort** [comment]
@kingjr I am very happy to see you growing up quickly wrt software design trade offs :)

**Comment by jona-sassenhagen** [comment]
IMO if you drop this, there should be an example showing how to implement e.g. regression.

**Comment by agramfort** [comment]
just to let you know it's possible to guess if we have a regressor in sklearn.

with latest sklearn we have is_regressor function. See

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/base.py#L456

**Comment by dengemann** [comment]
@kingjr I thought abou it, let's really keep it simple, it's good for the release. We'll cover examples in the notebook. For the next release we can see which APIs we want do add for scoring and classifiers.

**Comment by dengemann** [comment]
@kingjr are you happy here?

**Comment by kingjr** [comment]
@cmoutard Do you have time to try it off, and see whether it works on your usecases?

**Comment by cmoutard** [comment]
Unfortunately it is really unlikely I'll have time in the coming 10 days, sorry. But then, fully available to try it !

**Comment by dengemann** [comment]
Would be a pitty not to merge this before we release .... @cmoutard we might want to make a realease in the next days ...

**Comment by cmoutard** [comment]
I've just tested your branch @kingjr, and everything worked well from the first attempt (using the default predict_type).

**Comment by Eric89GXL** [comment]
More red than green makes me happy. @dengemann can you review and merge this one as you see fit? I'll go ahead and add the 0.9 tag and you can remove it if you feel it's inappropriate :)

**Comment by dengemann** [comment]
@kingjr thanks a lot!
